### PR TITLE
add support for SLOs

### PR DESCRIPTION
**Pull request by platinummonkey**
Adds support for Service Level Objectives

MA-434

### CODE REVIEW

File: datadog/dogshell/service_level_objective.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+            "--denominator",
+            help="denominator metric query (sum of total events)",
+            default=None,
+        )
+        create_parser.add_argument(
+            "--monitor_ids", help="explicit monitor_ids to use (CSV)", default=None
+        )
+        create_parser.add_argument(
+            "--monitor_search", help="monitor search terms to use", default=None
+        )
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: Is it missing the `groups` parameter for monitor SLOs ?

File: datadog/dogshell/service_level_objective.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+        create_parser.add_argument("--name", help="name of the SLO", default=None)
+        create_parser.add_argument(
+            "--description", help="description of the SLO", default=None
+        )
+        create_parser.add_argument(
+            "--tags", help="comma-separated list of tags", default=None
+        )
+        create_parser.add_argument(
+            "--thresholds",
+            help="comma separated list of <timeframe>:<target>[:<warning>]",
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: no `target_display` or `warning_display` ?

File: datadog/dogshell/service_level_objective.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        update_parser = verb_parsers.add_parser("update", help="Update existing SLO")
+        update_parser.add_argument(
+            "slo_id", help="SLO to replace with the new definition"
+        )
+        update_parser.add_argument(
+            "--type",
+            required=True,
+            help="type of the SLO (must specify it's original type)",
+            choices=["metric", "monitor"],
+        )
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: Why is this required if it mustn't change ?

File: datadog/dogshell/service_level_objective.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+
+        if args.type == "metric":
+            params["query"] = {
+                "numerator": args.numerator,
+                "denominator": args.denominator,
+            }
+        elif args.monitor_search:
+            params["monitor_search"] = args.monitor_search
+        else:
+            params["monitor_ids"] = args.monitor_ids
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: Should we add some logic to make sure we use the proper arguments with the proper SLO type ?

File: datadog/dogshell/service_level_objective.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+
+        if args.tags:
+            params["tags"] = args.tags.split(",")
+
+        res = api.ServiceLevelObjective.create(**params)
+        report_warnings(res)
+        report_errors(res)
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: `report_warnings` and `report_errors` won't work here, since the key in the response is `error`, not `errors`. And I don't know if you even return `warnings`

File: datadog/dogshell/service_level_objective.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+        else:
+            print(json.dumps(res))
+
+    @classmethod
+    def _file_create(cls, args):
+        api._timeout = args.timeout
+        format = args.format
+        slo = json.load(args.file)
+
+        params = {
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: Shouldn't we simply pass the loaded JSON ?

File: datadog/dogshell/service_level_objective.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+                "numerator": slo["numerator"],
+                "denominator": slo["denominator"],
+            }
+        elif slo.get("monitor_search"):
+            params["monitor_search"] = slo["monitor_search"]
+        else:
+            params["monitor_ids"] = slo["monitor_ids"]
+
+        if slo.get("tags"):
+            tags = sorted(set([t.strip() for t in slo["tags"].split(",") if t.strip()]))
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: In a file, this should already be a JSON list no ?

File: datadog/dogshell/service_level_objective.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+    def _file_update(cls, args):
+        api._timeout = args.timeout
+        format = args.format
+        slo = json.load(args.file)
+
+        params = {"type": slo["type"], "name": slo["name"]}
+
+        if slo.get("thresholds"):
+            thresholds = []
+            for threshold_str in args.thresholds.split(","):
```
Comment (zippolyte) [inline datadog/dogshell/service_level_objective.py]: same here, thresholds should already be in a proper JSON format, like you did in `file_create`

File: datadog/dogshell/service_level_objective.py (hunk 2 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: adding

File: datadog/dogshell/service_level_objective.py (hunk 1 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: adding 👍

File: datadog/dogshell/service_level_objective.py (hunk 3 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: this is required but the dogshell doesn't currently know what the current type is (though I could make it query and find out). this is more so it knows how to handle the update correctly.

File: datadog/dogshell/service_level_objective.py (hunk 4 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: adding some typing to argument parser

File: datadog/dogshell/service_level_objective.py (hunk 5 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: `warnings` is not currently returned, but keeping it consistent. `errors` is what is both reported and that function uses in the key

File: datadog/dogshell/service_level_objective.py (hunk 3 above)
Comment (zippolyte) [reply datadog/dogshell/service_level_objective.py]: I see, let's leave like this then

File: datadog/dogshell/service_level_objective.py (hunk 5 above)
Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: right now it throws an exception when there are errors, so wondering if this should be an arg to the service implementation 🤔

Comment (zippolyte) [reply datadog/dogshell/service_level_objective.py]: Actually yeah, you're right it raises an exception. I wonder how this even works in the other resources 🤔

Comment (zippolyte) [reply datadog/dogshell/service_level_objective.py]: Oh I see, the `APIClient.submit` method swallows the exceptions.
But in your case, since you're raising when the `error` key in the response payload is set, directly in you resource, it will not be swallowed by the super class.
Are you saying that you can have responses which contain `errors` plural, as well as the `error` singular that you use to raise exception ?

Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: yes :) in some cases the API will return actual exceptions, in other cases we return an `errors` which can contain 0 or more errors depending on the endpoint. for bulk delete, some can succeed while others not, their reasons for not succeeding will be in `errors`

typically the responses look like `{"data": ..., "errors": ...}` even for HTTP409 for example

Comment (platinummonkey) [reply datadog/dogshell/service_level_objective.py]: added a workaround method

File: datadog/api/api_client.py [hunk 9]
```diff
@@ -177,7 +184,7 @@ def submit(cls, method, path, api_version=None, body=None, attach_host_name=Fals
                 raise
         except ApiError as e:
             if _mute:
-                for error in e.args[0]['errors']:
+                for error in (e.args[0].get('errors') or []):
```
Comment (zippolyte) [inline datadog/api/api_client.py]:
```suggestion
                for error in (e.args[0].get('errors', [])):
```
